storage: omit unnecessary seek in `MVCCDelete`

erikgrinaker opened the pull request:
**roachpb: set `isRead` for `DeleteRequest`**

`DeleteRequest` performs a read to determine whether an existing, live
key was deleted, in order to populate the `DeleteResponse.FoundKey`
field. It must therefore have the `isRead` flag in order to avoid
deferring the `WriteTooOldError` and returning a possibly incorrect
value in the case of a write-write conflict (if the delete was submitted
below a newer key).

This has no externally visible effect on KV callers, because the
`WriteTooOldError` is retried on the client side (in
`txnSpanRefresher`), but it does violate the server contract. The flag
may also avoid an unnecessary Raft round trip to lay down an intent that
would be replaced later anyway.

Release note: None

**storage: omit unnecessary seek in `MVCCDelete`**

Recently, `MVCCDelete` was extended to return a boolean indicating
whether an existing, live key was deleted. This relied on passing a
`valueFn` callback to `mvccPutInternal()` which inspected the existing
value. Unfortunately, this incurs an additional, expensive seek. In the
common case when writing above the latest version, `mvccPutInternal`
has already read the existing value when synthesizing the meta record,
making this additional seek unnecessary.

This patch takes a conservative approach, by modifying `mvccPutInternal`
to return a boolean indicating whether an existing value was replaced by
the put. This is motivated by backport concerns. It does, however, incur
an additional read for all write-write conflicts, not just `MVCCDelete`,
but this cost is expected to be negligible considering the cost of a
transaction refresh.

A better solution might be to remove the need for an additional seek
with `valueFn` in the common case, which would also benefit e.g.
`MVCCConditionalPut` and `MVCCIncrement`, and avoid the read for
non-delete writes. This is left for the future.

```
name                     old time/op  new time/op  delta
MVCCPutDelete_Pebble-24  20.0µs ± 7%  15.5µs ± 3%  -22.43%  (p=0.008 n=5+5)

name                            old time/op  new time/op  delta
KV/Delete/Native/rows=1-24       153µs ± 6%   146µs ± 6%   -4.41%  (p=0.009 n=10+10)
KV/Delete/Native/rows=10-24      278µs ± 3%   227µs ± 4%  -18.22%  (p=0.000 n=10+10)
KV/Delete/Native/rows=100-24    1.28ms ± 3%  0.86ms ± 4%  -33.01%  (p=0.000 n=10+10)
KV/Delete/Native/rows=1000-24   10.8ms ± 4%   6.0ms ± 6%  -44.98%  (p=0.000 n=10+10)
KV/Delete/Native/rows=10000-24   110ms ± 4%    62ms ± 5%  -43.88%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=1-24          598µs ± 2%   588µs ± 2%   -1.62%  (p=0.009 n=10+10)
KV/Delete/SQL/rows=10-24         753µs ± 3%   701µs ± 2%   -6.90%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=100-24       2.34ms ± 2%  1.77ms ± 1%  -24.42%  (p=0.000 n=10+9)
KV/Delete/SQL/rows=1000-24      26.0ms ± 4%  18.5ms ± 1%  -29.08%  (p=0.000 n=10+9)
KV/Delete/SQL/rows=10000-24      248ms ± 8%   185ms ± 4%  -25.40%  (p=0.000 n=10+10)
```

Resolves #89253.

Release note: None

erikgrinaker committed:
storage: omit unnecessary seek in `MVCCDelete`

Recently, `MVCCDelete` was extended to return a boolean indicating
whether an existing, live key was deleted. This relied on passing a
`valueFn` callback to `mvccPutInternal()` which inspected the existing
value. Unfortunately, this incurs an additional, expensive seek. In the
common case when writing above the latest version, `mvccPutInternal`
has already read the existing value when synthesizing the meta record,
making this additional seek unnecessary.

This patch takes a conservative approach, by modifying `mvccPutInternal`
to return a boolean indicating whether an existing value was replaced by
the put. This is motivated by backport concerns. It does, however, incur
an additional read for all write-write conflicts, not just `MVCCDelete`,
but this cost is expected to be negligible considering the cost of a
transaction refresh.

A better solution might be to remove the need for an additional seek
with `valueFn` in the common case, which would also benefit e.g.
`MVCCConditionalPut` and `MVCCIncrement`, and avoid the read for
non-delete writes. This is left for the future.

```
name                     old time/op  new time/op  delta
MVCCPutDelete_Pebble-24  20.0µs ± 7%  15.5µs ± 3%  -22.43%  (p=0.008 n=5+5)

name                            old time/op  new time/op  delta
KV/Delete/Native/rows=1-24       153µs ± 6%   146µs ± 6%   -4.41%  (p=0.009 n=10+10)
KV/Delete/Native/rows=10-24      278µs ± 3%   227µs ± 4%  -18.22%  (p=0.000 n=10+10)
KV/Delete/Native/rows=100-24    1.28ms ± 3%  0.86ms ± 4%  -33.01%  (p=0.000 n=10+10)
KV/Delete/Native/rows=1000-24   10.8ms ± 4%   6.0ms ± 6%  -44.98%  (p=0.000 n=10+10)
KV/Delete/Native/rows=10000-24   110ms ± 4%    62ms ± 5%  -43.88%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=1-24          598µs ± 2%   588µs ± 2%   -1.62%  (p=0.009 n=10+10)
KV/Delete/SQL/rows=10-24         753µs ± 3%   701µs ± 2%   -6.90%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=100-24       2.34ms ± 2%  1.77ms ± 1%  -24.42%  (p=0.000 n=10+9)
KV/Delete/SQL/rows=1000-24      26.0ms ± 4%  18.5ms ± 1%  -29.08%  (p=0.000 n=10+9)
KV/Delete/SQL/rows=10000-24      248ms ± 8%   185ms ± 4%  -25.40%  (p=0.000 n=10+10)
```

Release note: None

erikgrinaker commented:
@nvanbenschoten Would you mind doublechecking real quick that I haven't missed anything in the intent rewrite cases?

nvb reviewed (commented):
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @nicktrav)

___
*[`pkg/storage/mvcc.go` line 1950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NDjXO0V9TXfhpuFNjsq:-NDjXO0V9TXfhpuFNjsr:bii37gc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e178caf27a6739d8a82b597c86d34b0aa3a1cf62/pkg/storage/mvcc.go#L1950)):*
> ```Go
> 				readTimestamp = writeTimestamp
> 			}
> 			if value, err = maybeGetValue(ctx, iter, key, value, ok, readTimestamp, valueFn); err != nil {
> ```

I'm struggling to page some of this logic back in, but I don't think this is correct for the transactional case where the txn's read timestamp lags below the timestamp of the most recently committed version. In that case, we need to continue serving the read portion of read-write requests (e.g. the `FoundKey` field on the `DeleteResponse`) using the read timestamp to ensure that the transaction operates on a consistent snapshot for its entire lifetime. So we can't always use the `meta.Deleted` field to populate the `FoundKey` field.

This raises an interesting point. The `FoundKey` field on `DeleteResponse` means that `DeleteRequest`s are no longer logically "blind". As a result, they should have the `isRead` flag and no longer qualify for `IsBlindWrite` (see the effect that has in `evaluateBatch`). However, I don't think we have a correctness bug today because the request's read doesn't impact what it writes or whether it throws an error (like it does for requests like ConditionalPut). Meanwhile, adding this flag would disqualify the request type from deferring WriteTooOld errors, which may be important for some workloads.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nicktrav, @nvanbenschoten, and @yuzefovich)

___
*[`pkg/storage/mvcc.go` line 1950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NDjXO0V9TXfhpuFNjsq:-NE0JkUg5h4yqOKGhjXR:b-dsz35n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e178caf27a6739d8a82b597c86d34b0aa3a1cf62/pkg/storage/mvcc.go#L1950)):*
>we need to continue serving the read portion of read-write requests (e.g. the FoundKey field on the DeleteResponse) using the read timestamp to ensure that the transaction operates on a consistent snapshot for its entire lifetime. So we can't always use the meta.Deleted field to populate the FoundKey field.

Right, even though we return a `WriteTooOldError` we may swallow that higher up if the transaction can successfully refresh its reads at the new timestamp and commit. That's fairly straightforward to fix, and I've added on a commit for it, but:

>This raises an interesting point. The FoundKey field on DeleteResponse means that DeleteRequests are no longer logically "blind". As a result, they should have the isRead flag and no longer qualify for IsBlindWrite (see the effect that has in evaluateBatch). However, I don't think we have a correctness bug today because the request's read doesn't impact what it writes or whether it throws an error (like it does for requests like ConditionalPut). Meanwhile, adding this flag would disqualify the request type from deferring WriteTooOld errors, which may be important for some workloads.

What if this affects the caller's logic? A delete below an existing key would return `FoundKey: false`, but might be "silently reevaluated" above the existing key (by e.g. the span refresher) without surfacing `FoundKey: true` to the caller. That seems like it could be problematic.

I don't think making this a non-blind write would necessarily cause a regression, since I believe SQL used to use `DeleteRange` for most deletes (?), which already set `isRead`.

@yuzefovich Any thoughts on how this would affect SQL callers?


<!-- Sent from Reviewable.io -->

yuzefovich commented:
> @yuzefovich Any thoughts on how this would affect SQL callers?

I'm on PTO for a couple of weeks (and only intermittently available during layovers :) ), so this response is based on a cursory look.

It seems like the only place where SQL looks at `kv.Result.Keys` (which is populated based on `FoundKey` boolean) is here:
https://github.com/cockroachdb/cockroach/blob/9e7e704180a0d72a55103a177204785ee462a5f3/pkg/sql/delete_range.go#L194-L209
and that logic is just counting the number of rows that were deleted. It sounds ok to me to under report this "row count" in rare cases like the one you mentioned.

erikgrinaker commented:
@nvanbenschoten Want to take another look at this? Marking this as a read seems like the correct approach here, but at least the current behavior is no worse than that prior to this PR. Then we can do a more thorough performance analysis for 23.1.

nvb reviewed (commented):
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nicktrav and @yuzefovich)

___
*[`pkg/storage/mvcc.go` line 1950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NDjXO0V9TXfhpuFNjsq:-NEb-vWR1k7j_4F-N9w1:bbasdxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e178caf27a6739d8a82b597c86d34b0aa3a1cf62/pkg/storage/mvcc.go#L1950)):*
> What if this affects the caller's logic? A delete below an existing key would return FoundKey: false, but might be "silently reevaluated" above the existing key (by e.g. the span refresher) without surfacing FoundKey: true to the caller. That seems like it could be problematic.

I'll try to recall / write-up how write-write conflict handling works in the hopes that it will guide us to the correct fix here.

There are two forms of "write too old" handling — deferred and eager.

Deferred write-too-old handling is possible when a KV write's result (the value in the intent that it writes) is not conditional on the value(s) that it reads. As a result, even if its write timestamp is pushed due to a write-write conflict, it can still go ahead and write an intent. It then sets the `Transaction.WriteTooOld` flag on the response transaction, which defers the need to refresh the transaction's read timestamp to some later time. A blind-write (e.g. `Put`) is the prototypical example of where deferred WTO handling is possible and helpful.

The eager write-too-old handling is meant for cases where a KV write's result is conditional on the value(s) that it reads. In these cases, a write-write conflict causes the write to fail. There's no use in performing the intent write based on the stale value(s) that were read if we know that they will change after a refresh/retry. Instead, we eagerly return a `WriteTooOld` error and eagerly refresh the transaction.

The lines between these forms of evaluation have been blurred by https://github.com/cockroachdb/cockroach/pull/38668 + https://github.com/cockroachdb/cockroach/pull/44654, but the split remains.

In either case, the read portion of the request (if any) should logically operate at the transaction's read timestamp so that it remains consistent with the rest of the reads that have been performed by the transaction. Those reads should also be tracked in the transaction's refresh spans so that any future refreshes (due to this write-write conflict or a future write-write conflict) validate that the reads are still valid. However, we avoid marking certain read-write requests as `needsRefresh` if two conditions both hold:
1. they perform eager write-too-old handling
2. their intent acts as a lock and prevents any outgoing read-write anti-dependencies (i.e. the refresh of their read portion to fail) on their read set from forming after the fact.

If both of these conditions are true, then there is no need to refresh the read set of of the read-write request and any read that they performed will remain consistent with the rest of their txn.

This leads us to the conclusion that if we don't want `DeleteRequest` to need to refresh, we should have it perform eager write-too-old handling by marking it as `isRead`. If we do that then we can/should have its `FoundKey` response field set to the value of the condition at the txn's write timestamp, not its read timestamp. So that would allow what you have here to be correct.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nicktrav, @nvanbenschoten, and @yuzefovich)

___
*[`pkg/storage/mvcc.go` line 1950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NDjXO0V9TXfhpuFNjsq:-NEjcfsMAuHc-mR791te:bea4uku) ([raw file](https://github.com/cockroachdb/cockroach/blob/e178caf27a6739d8a82b597c86d34b0aa3a1cf62/pkg/storage/mvcc.go#L1950)):*
Thanks for the background @nvanbenschoten, this is really helpful. I'll need to read the relevant code more closely, but I've updated `DeleteRequest` with `isRead` but not `needsRefresh`. I've also removed the additional commit that attempted to evaluate `FoundKey` at the read timestamp, instead evaluating at the write timestamp, which I believe is equivalent to the meta timestamp for our purposes here.

> There's no use in performing the intent write based on the stale value(s) that were read if we know that they will change after a refresh/retry. Instead, we eagerly return a WriteTooOld error and eagerly refresh the transaction.

`mvccPutInternal` will currently still lay down an intent for the tombstone on write-write conflicts. This appears to be unnecessary given the above, but I'd rather not make too many changes to the logic here for the 22.2 backport, and from what I understand this seems to be solely a performance issue.

>So that would allow what you have here to be correct.

By "this", I'm assuming you meant the initial version (i.e. initial commit), so I've removed the second commit but kept the testing improvements.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nicktrav, @nvanbenschoten, and @yuzefovich)

___
*[`pkg/storage/mvcc.go` line 1950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NDjXO0V9TXfhpuFNjsq:-NEk6iL-3VCmERRl7cpM:b-oqyba8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e178caf27a6739d8a82b597c86d34b0aa3a1cf62/pkg/storage/mvcc.go#L1950)):*
>>There's no use in performing the intent write based on the stale value(s) that were read if we know that they will change after a refresh/retry. Instead, we eagerly return a WriteTooOld error and eagerly refresh the transaction.
>
>mvccPutInternal will currently still lay down an intent for the tombstone on write-write conflicts. This appears to be unnecessary given the above, but I'd rather not make too many changes to the logic here for the 22.2 backport, and from what I understand this seems to be solely a performance issue.

I may have misunderstood the above. So the idea here is that we'll lay down the intent at the WriteTooOld timestamp, returning FoundKey as of that timestamp, the transaction will get pushed to that timestamp, and we won't need to refresh the write because both the write and the FoundKey read are consistent as of that timestamp? Is that how this will also play out in practice?


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 5 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-:-NEm3F-SCpKIlde8otpw:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @nicktrav, and @yuzefovich)

___
*[`pkg/storage/mvcc.go` line 1950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NDjXO0V9TXfhpuFNjsq:-NEm0GuS7wsilbJSj0l7:b9vfimi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e178caf27a6739d8a82b597c86d34b0aa3a1cf62/pkg/storage/mvcc.go#L1950)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

>>There's no use in performing the intent write based on the stale value(s) that were read if we know that they will change after a refresh/retry. Instead, we eagerly return a WriteTooOld error and eagerly refresh the transaction.
>
>mvccPutInternal will currently still lay down an intent for the tombstone on write-write conflicts. This appears to be unnecessary given the above, but I'd rather not make too many changes to the logic here for the 22.2 backport, and from what I understand this seems to be solely a performance issue.

I may have misunderstood the above. So the idea here is that we'll lay down the intent at the WriteTooOld timestamp, returning FoundKey as of that timestamp, the transaction will get pushed to that timestamp, and we won't need to refresh the write because both the write and the FoundKey read are consistent as of that timestamp? Is that how this will also play out in practice?
</blockquote></details>

If we add the `isRead` flag to the `DeleteRequest`, then it will no longer qualify as a blind write. That means that it will not be able to defer WriteTooOld errors returned during evaluation, so any intent write that it adds to the WriteBatch will be discarded on a write-write conflict, along with the BatchResponse.

In principle, MVCC logic should still be performing reads at the transaction's read timestamp (not the meta timestamp). However, in practice, if a WriteTooOld error can't be deferred and we know that a write-write conflict implies a discarded response then it shouldn't matter what timestamp the information on the response comes from.

Note that this is not the case for operations like CPut, where the read can have side effects (throw a ConditionFailedError), so we need to ensure that it's performed at the transaction's read timestamp.
___
*[`pkg/storage/mvcc.go` line 1811 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NEm2GjLEOdHdRcYyYSx:-NEm2GjLEOdHdRcYyYSy:b-mwgc9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08baf527bfacc18cfb1ffb6a6af2d74b1e54423/pkg/storage/mvcc.go#L1811)):*
> ```Go
> 			}
> 
> 			exReplaced = exVal.exists && exVal.IsPresent()
> ```

nit: doesn't `optionalValue.IsPresent` already check the `exists` flag?
___
*[`pkg/storage/testdata/mvcc_histories/deletes` line 90 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NEm2r55E9RaGRWgKCx6:-NEm2r55E9RaGRWgKCx7:b-ij9oq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08baf527bfacc18cfb1ffb6a6af2d74b1e54423/pkg/storage/testdata/mvcc_histories/deletes#L90)):*
> ```
> # Try deleting historical versions using a txn with an old read timestamp, and
> # ensure "found key" is reported correctly for various timestamps along with the
> # WriteTooOldError (the latter may be swallowed by the transaction coordinator).
> ```

s/the transaction coordinator/request evaluation logic/


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
@erikgrinaker given how subtle this area appears to be, should we discuss whether the regression that this resolves is large enough to risk an immediate backport of this PR? 4.5µs is not negligible for deletes, but I'm also not aware of a workload that was already using `MVCCDelete` (as opposed to `MVCCDeleteRange`) and would also be sensitive enough to this to notice. Are you? Perhaps Raft log truncation?

One lower-risk change we could make is to add an option to `MVCCDelete` that lets a caller request the `foundKey` return value. If not requested, the `valueFn` is not created and provided to `mvccPutUsingIter`. That way, only `DeleteRequest` pays the cost and other callers of `MVCCDelete` are not impacted. If we did that, we would still need to add the `isRead` flag to `DeleteRequest`.

erikgrinaker reviewed (commented):
> should we discuss whether the regression that this resolves is large enough to risk an immediate backport of this PR? 4.5µs is not negligible for deletes, but I'm also not aware of a workload that was already using MVCCDelete (as opposed to MVCCDeleteRange) and would also be sensitive enough to this to notice. Are you? Perhaps Raft log truncation?

Plain Raft log truncation isn't affected, since it uses engine clears directly. `MVCCDelete` is only used when truncating an uncommitted Raft log tail during appends, which seems fine (and this could really just use engine clears directly as well). The other callers are cleanups of transaction records and abort span entries, both of which are inline deletes and likely not performance sensitive enough to matter.

I think the `isRead` correctness issue is the bigger problem here, and that already applies to `release-22.2`, so we'll need to do something about that. I think we can probably get away with the `valueFn` performance regression since `MVCCDeleteRange` certainly didn't do any better, and Tobi also pointed out that this won't matter much considering we're also doing a Raft roundtrip. So I'm ok with keeping the unconditional `valueFn` instead, but do you think we need to? I suppose the surrounding logic has already been hardened for the `valueFn` case, so it's the conservative option, but it isn't clear to me that the end result differs significantly from the current PR logic.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-:-NEmA4X603CnZZ3cAy9o:b-r2roat)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nicktrav, @nvanbenschoten, and @yuzefovich)

___
*[`pkg/storage/mvcc.go` line 1950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NDjXO0V9TXfhpuFNjsq:-NEm6AnM5fcT8FpE1tcp:b-55ssk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e178caf27a6739d8a82b597c86d34b0aa3a1cf62/pkg/storage/mvcc.go#L1950)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we add the `isRead` flag to the `DeleteRequest`, then it will no longer qualify as a blind write. That means that it will not be able to defer WriteTooOld errors returned during evaluation, so any intent write that it adds to the WriteBatch will be discarded on a write-write conflict, along with the BatchResponse.

In principle, MVCC logic should still be performing reads at the transaction's read timestamp (not the meta timestamp). However, in practice, if a WriteTooOld error can't be deferred and we know that a write-write conflict implies a discarded response then it shouldn't matter what timestamp the information on the response comes from.

Note that this is not the case for operations like CPut, where the read can have side effects (throw a ConditionFailedError), so we need to ensure that it's performed at the transaction's read timestamp.
</blockquote></details>

I see, that's what I initially thought (except I didn't consider that the batch would be discarded). It's unfortunate that we'll have to reevaluate the write at the higher timestamp though.
___
*[`pkg/storage/mvcc.go` line 1811 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-NEm2GjLEOdHdRcYyYSx:-NEm8Ozf-6wERI2oJ0vs:bo6apac) ([raw file](https://github.com/cockroachdb/cockroach/blob/d08baf527bfacc18cfb1ffb6a6af2d74b1e54423/pkg/storage/mvcc.go#L1811)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: doesn't `optionalValue.IsPresent` already check the `exists` flag?
</blockquote></details>

Indeed.


<!-- Sent from Reviewable.io -->

erikgrinaker committed:
roachpb: set `isRead` for `DeleteRequest`

`DeleteRequest` performs a read to determine whether an existing, live
key was deleted, in order to populate the `DeleteResponse.FoundKey`
field. It must therefore have the `isRead` flag in order to avoid
deferring the `WriteTooOldError` and returning a possibly incorrect
value in the case of a write-write conflict (if the delete was submitted
below a newer key).

This has no externally visible effect on KV callers, because the
`WriteTooOldError` is retried on the client side (in
`txnSpanRefresher`), but it does violate the server contract. The flag
may also avoid an unnecessary Raft round trip to lay down an intent that
would be replaced later anyway.

Release note: None

erikgrinaker reviewed (commented):
Decided to split out the `isRead` issue to a separate PR. However, as I was writing a test case for the txn handling, it seems like the `DeleteRequest` is always retried at the higher write timestamp, thus returning the correct `FoundKey: true`, even without `isRead`. I'll have to dig into why that's happening, but I'm going to put this on the backburner for a bit and look into a different release blocker.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-:-NEp2EHH8Hy2QbGVWCTK:betast8)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nicktrav, @nvanbenschoten, and @yuzefovich)



<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
Yeah, I looked a bit closer at this, and without `isRead` the `WriteTooOldError` causes an immediate server-side refresh and retry at the newer timestamp in `evaluateWriteBatchWithServersideRefreshes`. If the txn has already done a read, the txn span refresher performs a refresh and retry at the newer timestamp. With `isRead`, we're missing out on the server-side refresh, but it appears to already give the correct result so it's not clear to me why we need it. Can you give a history that would produce an inconsistency here?

Simple test case, for reference:

```go
func TestDeleteFoundKeyWriteTooOld(t *testing.T) {
	defer leaktest.AfterTest(t)()
	defer log.Scope(t).Close(t)

	ctx := context.Background()
	tr := tracing.NewTracer()
	recCtx, getRecAndFinish := tracing.ContextWithRecordingSpan(ctx, tr, "test")
	defer getRecAndFinish()

	s, _, db := serverutils.StartServer(t, base.TestServerArgs{Tracer: tr})
	defer s.Stopper().Stop(ctx)

	txn := db.NewTxn(ctx, "test")
	//_, err := txn.Get(ctx, "a")
	//require.NoError(t, err)
	require.NoError(t, txn.Put(ctx, "a", "a")) // replace with Get for txnSpanRefresher
	require.NoError(t, db.Put(ctx, "b", "b"))

	var ba roachpb.BatchRequest
	ba.Add(&roachpb.DeleteRequest{
		RequestHeader: roachpb.RequestHeader{
			Key: roachpb.Key("b"),
		},
	})
	br, pErr := txn.Send(recCtx, ba)
	require.NoError(t, pErr.GoError())
	require.True(t, br.Responses[0].GetDelete().FoundKey)
	require.NoError(t, txn.Commit(ctx))

	fmt.Println(getRecAndFinish())
}
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-:-NEpqCHC0d-NZQOwL4Se:b-c0i0t7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nicktrav, @nvanbenschoten, and @yuzefovich)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
FWIW, I also tried disabling the server-side refreshes entirely, in the hope that it would reveal any issues with the `txnSpanRefresher`, but the delete is still refreshed at the higher timestamp, returning the correct `FoundKey`.

erikgrinaker commented:
Ran some more benchmarks. `workload run kv` benchmarks with `--del-percent` were inconclusive (too noisy). However, SQL KV benchmarks showed a minor improvement for this PR:

```
name                            old time/op  new time/op  delta
KV/Delete/Native/rows=1-24       139µs ± 3%   139µs ± 5%    ~     (p=0.393 n=10+10)
KV/Delete/Native/rows=10-24      209µs ± 5%   208µs ± 3%    ~     (p=0.971 n=10+10)
KV/Delete/Native/rows=100-24     706µs ± 2%   701µs ± 2%    ~     (p=0.353 n=10+10)
KV/Delete/Native/rows=1000-24   5.10ms ± 1%  4.95ms ± 2%  -2.94%  (p=0.000 n=8+10)
KV/Delete/Native/rows=10000-24  53.2ms ± 3%  51.0ms ± 2%  -4.06%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=1-24          612µs ± 2%   600µs ± 2%  -1.85%  (p=0.028 n=9+10)
KV/Delete/SQL/rows=10-24         708µs ± 2%   711µs ± 2%    ~     (p=0.529 n=10+10)
KV/Delete/SQL/rows=100-24       1.43ms ± 2%  1.41ms ± 2%    ~     (p=0.123 n=10+10)
KV/Delete/SQL/rows=1000-24      3.11ms ± 3%  3.10ms ± 2%    ~     (p=0.684 n=10+10)
KV/Delete/SQL/rows=10000-24     26.4ms ± 2%  26.6ms ± 4%    ~     (p=0.315 n=10+10)
```

For comparison, here's the comparison between 22.1 and `master`:

```
name                            old time/op  new time/op  delta
KV/Delete/Native/rows=1-24       135µs ± 5%   139µs ± 3%     ~     (p=0.075 n=10+10)
KV/Delete/Native/rows=10-24      202µs ± 4%   209µs ± 5%   +3.30%  (p=0.011 n=10+10)
KV/Delete/Native/rows=100-24     692µs ± 3%   706µs ± 2%   +2.04%  (p=0.015 n=10+10)
KV/Delete/Native/rows=1000-24   5.03ms ± 1%  5.10ms ± 1%   +1.43%  (p=0.000 n=9+8)
KV/Delete/Native/rows=10000-24  51.5ms ± 2%  53.2ms ± 3%   +3.20%  (p=0.004 n=10+10)
KV/Delete/SQL/rows=1-24          493µs ± 4%   612µs ± 2%  +24.17%  (p=0.000 n=10+9)
KV/Delete/SQL/rows=10-24         624µs ± 3%   708µs ± 2%  +13.46%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=100-24       1.77ms ± 4%  1.43ms ± 2%  -19.55%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=1000-24      3.04ms ± 4%  3.11ms ± 3%   +2.36%  (p=0.023 n=10+10)
KV/Delete/SQL/rows=10000-24     24.7ms ± 3%  26.4ms ± 2%   +6.66%  (p=0.000 n=10+10)
```

Comparing 22.1 to this PR show that it addresses much of the regression for large batches, but there's still a significant regression in the constant cost of small batches:

```
name                            old time/op  new time/op  delta
KV/Delete/Native/rows=1-24       135µs ± 5%   139µs ± 5%     ~     (p=0.165 n=10+10)
KV/Delete/Native/rows=10-24      202µs ± 4%   208µs ± 3%   +2.91%  (p=0.019 n=10+10)
KV/Delete/Native/rows=100-24     692µs ± 3%   701µs ± 2%     ~     (p=0.063 n=10+10)
KV/Delete/Native/rows=1000-24   5.03ms ± 1%  4.95ms ± 2%   -1.55%  (p=0.001 n=9+10)
KV/Delete/Native/rows=10000-24  51.5ms ± 2%  51.0ms ± 2%     ~     (p=0.218 n=10+10)
KV/Delete/SQL/rows=1-24          493µs ± 4%   600µs ± 2%  +21.88%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=10-24         624µs ± 3%   711µs ± 2%  +13.87%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=100-24       1.77ms ± 4%  1.41ms ± 2%  -20.20%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=1000-24      3.04ms ± 4%  3.10ms ± 2%     ~     (p=0.052 n=10+10)
KV/Delete/SQL/rows=10000-24     24.7ms ± 3%  26.6ms ± 4%   +7.70%  (p=0.000 n=10+10)
```

> I'm also not aware of a workload that was already using MVCCDelete (as opposed to MVCCDeleteRange) and would also be sensitive enough to this to notice

Note that while there may not be many `MVCCDelete` callers, there's a significant number of `DeleteRequest` callers that will be affected, such as index backfills (due to the merge step of the incremental index backfiller).

nvanbenschoten commented:
> I also tried disabling the server-side refreshes entirely, in the hope that it would reveal any issues with the txnSpanRefresher, but the delete is still refreshed at the higher timestamp, returning the correct FoundKey.

I think what you're seeing is that we currently [immediately refresh and retry](https://github.com/cockroachdb/cockroach/blob/c8fc390ac97bfe5ce2735151180d4bd85652e2da/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L234) requests (from the client) on "deferred" write too old errors. This is what I meant by the lines between eager and deferred WriteTooOld handling being "blurred".

erikgrinaker commented:
I met with Nathan yesterday to clarify a few aspects. Quick summary.

* The `isRead` flag has no externally visible effect, because the `WriteTooOldError` will be retried on the server side (`evaluateWriteBatchWithServersideRefreshes`) or client side (`txnSpanRefresher`).

* The `isRead` flag is still desirable, because it may avoid an unnecessary Raft roundtrip to lay down an intent that we'll shortly be replacing anyway (due to the deferred WTO handling).

* We can maintain serializability and also optimize the common case by using the meta record when we're reading above the meta timestamp, and only do an additional read on write-write conflicts (which will be expensive anyway).

Will update the PR shortly.

erikgrinaker commented:
This should be ready for a final look, @nvanbenschoten.

erikgrinaker commented:
I also ran some `KV/Delete` benchmarks on top of #90563. While that doesn't account for network and fsync latency, it still demonstrates a clear benefit when deleting many keys in the O(10ms) range:

```
name                            old time/op  new time/op  delta
KV/Delete/Native/rows=1-24       153µs ± 6%   146µs ± 6%   -4.41%  (p=0.009 n=10+10)
KV/Delete/Native/rows=10-24      278µs ± 3%   227µs ± 4%  -18.22%  (p=0.000 n=10+10)
KV/Delete/Native/rows=100-24    1.28ms ± 3%  0.86ms ± 4%  -33.01%  (p=0.000 n=10+10)
KV/Delete/Native/rows=1000-24   10.8ms ± 4%   6.0ms ± 6%  -44.98%  (p=0.000 n=10+10)
KV/Delete/Native/rows=10000-24   110ms ± 4%    62ms ± 5%  -43.88%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=1-24          598µs ± 2%   588µs ± 2%   -1.62%  (p=0.009 n=10+10)
KV/Delete/SQL/rows=10-24         753µs ± 3%   701µs ± 2%   -6.90%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=100-24       2.34ms ± 2%  1.77ms ± 1%  -24.42%  (p=0.000 n=10+9)
KV/Delete/SQL/rows=1000-24      26.0ms ± 4%  18.5ms ± 1%  -29.08%  (p=0.000 n=10+9)
KV/Delete/SQL/rows=10000-24      248ms ± 8%   185ms ± 4%  -25.40%  (p=0.000 n=10+10)
```

22.1 compared with this PR also shows a clear benefit due to the SQL-level optimizations, but a minor regression for native KV deletes that's consistent with known MVCC range tombstone overhead:

```
name                            old time/op  new time/op  delta
KV/Delete/Native/rows=1-24       138µs ± 4%   146µs ± 6%   +5.90%  (p=0.000 n=10+10)
KV/Delete/Native/rows=10-24      218µs ± 5%   227µs ± 4%   +4.25%  (p=0.000 n=10+10)
KV/Delete/Native/rows=100-24     800µs ± 6%   856µs ± 4%   +6.89%  (p=0.000 n=10+10)
KV/Delete/Native/rows=1000-24   5.81ms ± 4%  5.97ms ± 6%     ~     (p=0.089 n=10+10)
KV/Delete/Native/rows=10000-24  58.5ms ± 6%  61.6ms ± 5%   +5.39%  (p=0.005 n=10+10)
KV/Delete/SQL/rows=1-24          586µs ± 3%   588µs ± 2%     ~     (p=0.393 n=10+10)
KV/Delete/SQL/rows=10-24         807µs ± 5%   701µs ± 2%  -13.18%  (p=0.000 n=10+10)
KV/Delete/SQL/rows=100-24       2.76ms ± 3%  1.77ms ± 1%  -35.98%  (p=0.000 n=10+9)
KV/Delete/SQL/rows=1000-24      33.1ms ± 2%  18.5ms ± 1%  -44.19%  (p=0.000 n=10+9)
KV/Delete/SQL/rows=10000-24      661ms ± 3%   185ms ± 4%  -72.03%  (p=0.000 n=10+10)
```

nvb reviewed (approved):
I like it! And thanks for filing https://github.com/cockroachdb/cockroach/issues/90609. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 5 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89375#-:-NFGM2NB5IvkSWfjDymD:b-orytp4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nicktrav)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
TFTR!

bors r+
